server/pkg/controller/embedding/controller.go

```diff
@@ -368,32 +395,99 @@ func (c *Controller) getEmbeddingObjectsParallelV2(userID int64, dbEmbeddingRows
 	return embeddingObjects, nil
 }
 
-func (c *Controller) getEmbeddingObject(ctx context.Context, objectKey string, downloader *s3manager.Downloader) (ente.EmbeddingObject, error) {
-	return c.getEmbeddingObjectWithRetries(ctx, objectKey, downloader, 3)
+type getOptions struct {
+	RetryCount   int
+	FetchTimeOut gTime.Duration
 }
 
-func (c *Controller) getEmbeddingObjectWithRetries(ctx context.Context, objectKey string, downloader *s3manager.Downloader, retryCount int) (ente.EmbeddingObject, error) {
+func (c *Controller) getEmbeddingObject(ctx context.Context, objectKey string, downloader *s3manager.Downloader, opt *getOptions) (ente.EmbeddingObject, error) {
+	if opt == nil {
+		opt = &getOptions{
+			RetryCount:   3,
+			FetchTimeOut: embeddingFetchTimeout,
+		}
+	}
+	ctxLogger := log.WithField("objectKey", objectKey)
+	totalAttempts := opt.RetryCount + 1
+	for i := 0; i < totalAttempts; i++ {
+		// Create a new context with a timeout for each fetch
+		fetchCtx, cancel := context.WithTimeout(ctx, opt.FetchTimeOut)
+		select {
+		case <-ctx.Done():
+			cancel()
+			return ente.EmbeddingObject{}, stacktrace.Propagate(ctx.Err(), "")
+		default:
+			obj, err := c.downloadObject(fetchCtx, objectKey, downloader, c.embeddingBucket)
+			cancel() // Ensure cancel is called to release resources
+			if err == nil {
+				return obj, nil
+			}
+			// Check if the error is due to context timeout or cancellation
+			if err == nil && fetchCtx.Err() != nil {
+				ctxLogger.Error("Fetch timed out or cancelled: ", fetchCtx.Err())
+			} else {
+				// check if the error is due to object not found
+				if s3Err, ok := err.(awserr.RequestFailure); ok {
+					if s3Err.Code() == s3.ErrCodeNoSuchKey {
+						if c.areEmbeddingAndHotBucketSame {
+							ctxLogger.Error("Object not found: ", s3Err)
```

Comment:
We should abort instead of retrying when this happens (Not fully aware of how the go control flow is working here, so please ignore if this is already what the code is doing)